drivers: digital-io: max22196 : Add driver support for MAX22196

RaduSabau1 opened the pull request:
## Pull Request Description

The MAX22196 is an industrial octal digital input that translates eight industrial 24V or TTL level inputs to logic level outputs. The device has a serial interface allowing configuration and reading of serialized data through SPI.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [ ] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [ ] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [x] I have updated the documentation (wiki pages, ReadMe etc), if applies

amiclaus commented on drivers/digital-io/max22196/README.rst:
```diff
@@ -0,0 +1,147 @@
+MAX22196 no-OS driver
```
please link in this commit also the readme file to the drivers ToC (after #1955 gets merged, to avoid possible conflicts)

RaduSabau1 commented:
V2

- Included README in sphinx/source.

rbolboac commented on drivers/digital-io/max22196/max22196.c:
```diff
@@ -0,0 +1,424 @@
+
+/**
+ * @brief Compute the CRC5 value for an array of bytes when writing to MAX22196
+ * @param data - array of data to encode
```
missing param description for encode

rbolboac commented on drivers/digital-io/max22196/max22196.c:
```diff
@@ -0,0 +1,424 @@
+	no_os_spi_remove(desc->comm_desc);
+
+	no_os_free(desc);
+
```
missing gpio remove

rbolboac commented on drivers/digital-io/max22196/max22196.h:
```diff
@@ -0,0 +1,141 @@
+	bool crc_en;
+};
+
+int max22196_reg_write(struct max22196_desc *, uint32_t, uint32_t);
+
+int max22196_reg_read(struct max22196_desc *, uint32_t, uint32_t *);
+
+int max22196_reg_update(struct max22196_desc *, uint32_t, uint32_t, uint32_t);
+
+int max22196_set_mode(struct max22196_desc *, uint32_t, enum max22196_mode);
+
+int max22196_chan_cfg(struct max22196_desc *, uint32_t, uint32_t,
+		      enum max22196_curr);
+
+int max22196_filter_set(struct max22196_desc *, uint32_t, uint32_t,
+			enum max22196_delay);
+
+int max22196_filter_get(struct max22196_desc *, uint32_t, uint32_t *,
+			enum max22196_delay *);
+
+int max22196_set_chan_cnt(struct max22196_desc *, uint32_t, uint8_t, uint8_t);
+
+int max22196_get_chan_cnt(struct max22196_desc *, uint32_t, uint8_t *,
+			  uint8_t *);
+
+int max22196_init(struct max22196_desc **, struct max22196_init_param *);
+
+int max22196_remove(struct max22196_desc *);
```
missing APIs brief description, I see in max22190 there are present so would be nice to have them here as well

RaduSabau1 commented:
V3

- Added APIs brief description
- Addressed other comments.

CiprianRegus commented on drivers/digital-io/max22196/iio_max22196.c:
```diff
@@ -0,0 +1,695 @@
+
+	descriptor->iio_dev = &max22196_iio_dev;
+
+	ret = max22196_iio_setup_channels(descriptor,
```
All your channels are input. Why wouldn't you enable them all?

RaduSabau1 replied:
Will remove chans_enabled boolean since it's not needed in this case and will setup all the channels to be enabled.

CiprianRegus commented on drivers/digital-io/max22196/max22196.c:
```diff
@@ -0,0 +1,428 @@
+	if (ret)
+		goto spi_error;
+
+	if (descriptor->crc_desc) {
+		ret = no_os_gpio_get_value(descriptor->crc_desc, &reg_val);
+		if (ret)
+			goto gpio_error;
+		if (reg_val == NO_OS_GPIO_HIGH)
+			descriptor->crc_en = true;
+		else
+			descriptor->crc_en = false;
+	}
```
This may also be set by the target microcontroller (as in the CRC pin being output), so the logic won't work in such case.

RaduSabau1 replied:
I will change the code so CRC GPIO needs to be set as an output from the microcontroller (since it's an input in the case of the MAX22196) and check the value afterwards.

CiprianRegus commented on drivers/digital-io/max22196/max22196.c:
```diff
@@ -0,0 +1,428 @@
+ * @param cnt_lsb_byte - Variable through which the counter's lsb byte is read.
+ * @return 0 in case of succes, negative error code otherwise.
+*/
+int max22196_get_chan_cnt(struct max22196_desc *desc, uint32_t ch,
```
just use a 16 bit variable...

CiprianRegus commented on drivers/digital-io/max22196/max22196.c:
```diff
@@ -0,0 +1,428 @@
+ * @param cnt_lsb_byte - Counter value to be set for the LSB Byte.
+ * @return 0 in case of succes, negative error code otherwise.
+*/
+int max22196_set_chan_cnt(struct max22196_desc *desc, uint32_t ch,
```
same as in the get() case.

RaduSabau1 commented on drivers/digital-io/max22196/README.rst:
```diff
@@ -0,0 +1,147 @@
+Device Attributes
+-----------------
+
+MAX22196 has a total of 10 attributes :
```
Fixing it.

RaduSabau1 commented:
V4

- CRC GPIO is now required to be output if set.
- Changed the API for counter to use a uint16_t variable instead of two uint8_t variables.
- All channels in the IIO driver are now enabled, removed the boolean array in the iio_desc_init_param.
- Fixed README documentation.

RaduSabau1 commented:
V5

- Fixed README.

RaduSabau1 commented:
V6

- Solved error in init function regarding set_chan_cnt function call.

rbolboac commented on drivers/digital-io/max22196/iio_max22196.c:
```diff
@@ -0,0 +1,682 @@
+	struct iio_channel *max22196_iio_channels;
+	uint32_t i;
+
+	max22196_iio_channels = no_os_calloc(MAX22196_CHANNELS,
+					     sizeof(*max22196_iio_channels));
```
there is no need to allocate these channels dynamically seeing how they do not change function or number or anything else

RaduSabau1 commented:
V7

- Added API brief description and structure brief description.
- Fixed warnings in driver and IIO driver.
- max22196_iio_remove now returns 0

rbolboac commented on drivers/digital-io/max22196/max22196.c:
```diff
@@ -0,0 +1,528 @@
+ * @param delay - Filter delay to be set.
+ * @return 0 in case of succes, negative error code otherwise.
+*/
+int max22196_filter_set(struct max22196_desc *desc, uint32_t ch,
```
what about CLRFILTR which is present in max22190?

rbolboac commented on drivers/digital-io/max22196/max22196.c:
```diff
@@ -0,0 +1,528 @@
+ * @param enabled - State of the specific configuration.
+ * @return 0 in case of succes, negative error code otherwise.
+*/
+int max22196_global_cfg(struct max22196_desc *desc,
```
in max22190 REFDISHTCFG is set in fault_set... I think we should have the same handling in both drivers. same with clrfiltr

rbolboac commented on drivers/digital-io/max22196/max22196.c:
```diff
@@ -0,0 +1,557 @@
+					   MAX22196_FAULT_MASK(fault_mask),
+					   no_os_field_prep(MAX22196_FAULT_MASK(fault_mask),
+							   enabled));
+	case MAX22196_FAULT1_VMLOW ... MAX22196_FAULT1_FAULT2:
```
the logic from max22190 should be followed here as well,  regarding fault2 enable bit in fault1 register

rbolboac commented on drivers/digital-io/max22196/max22196.c:
```diff
@@ -0,0 +1,557 @@
+
+	no_os_spi_remove(desc->comm_desc);
+
+	if (desc->crc_desc)
```
I don't think the if condition is necessary

rbolboac commented on drivers/digital-io/max22196/max22196.c:
```diff
@@ -0,0 +1,557 @@
+		return -ENODEV;
+
+	for (i = 0; i < MAX22196_CHANNELS; i++) {
+		ret = max22196_set_chan_cnt(desc, i, 0);
```
in max22190 you are using a macro instead of 0 value, let's try to keep things as similar as possible between the two driver

RaduSabau1 commented:
V8

- Applied same logic for fault2en and using it as a bit mask.
- Removed unnecessary lines.
- Addressed other comments.

RaduSabau1 commented:
V9

- Updated README.rst documentation.
